### PR TITLE
add stronger types to ReportProvider and related components

### PR DESCRIPTION
**Pull request by braxex**
## Description

Our ReportProvider and many report- and reportData- related components were lacking strong types. This PR adds those types and updates all downstream components that were relying on (read: exploiting the weakness of) the previous `any` types. 

It was also necessary to update test files and global report- and reportData- related mocks (in `setupJest.tsx`) that were in a similar situation (aligned to the previous weak types). Many of the test files were unnecessarily creating their own custom mocks too -- those have been deleted and replaced with the global mocks, which provide enough flexibility for our test cases.

I also took the opportunity to fix some `react-router` vs. `react-router-dom` imports and mocking (we should always be using `react-router-dom` instead of `react-router` for these).

Bon appétit.

### How to test

- [x] All the tests pass

### Changed Dependencies

n/a

## Code author checklist
- [x] I have performed a self-review of my code
- [x] I have added [thorough](https://bit.ly/3zPrxuZ) tests
- [x] ~~I have added analytics, if necessary~~
- [x] ~~I have updated the documentation, if necessary~~

## Reviewer checklist (two different people)
- [ ] I have done the deep review and verified the items in the above checklist are g2g
- [ ] I have done the lgtm review

### CODE REVIEW

File: services/ui-src/src/components/reports/ReportPage.test.tsx [hunk 1]
```diff
@@ -56,77 +27,97 @@ jest.mock("utils", () => ({
...
-const standardReportPageComponent = (
+const mockUpdateReport = jest.fn();
+const mockUpdateReportData = jest.fn();
+const mockedReportContext = {
+  ...mockReportContext,
+  updateReport: mockUpdateReport,
+  updateReportData: mockUpdateReportData,
+};
+
+const ReportPageComponent_StandardForm = (
```
Comment (ntsummers1) [inline services/ui-src/src/components/reports/ReportPage.test.tsx:38]: Could we use the abstracted code you wrote here? Specifically these lines and use them as the mocked report context?

```
export const mockReportMethods = {
  setReport: jest.fn(),
  fetchReport: jest.fn(),
  updateReport: jest.fn(),
  removeReport: jest.fn(),
  setReportData: jest.fn(),
  fetchReportData: jest.fn(),
  updateReportData: jest.fn(),
  fetchReportsByState: jest.fn(),
};

export const mockReportContext = {
  ...mockReportMethods,
  report: mockReport,
  reportData: mockReportData,
  reportsByState: mockReportsByState,
  errorMessage: "",
};
```

File: services/ui-src/src/routes/ReviewSubmit/ReviewSubmit.test.tsx [hunk 2]
```diff
@@ -6,84 +6,73 @@ import { ReviewSubmit } from "routes";
...
   },
 }));
 
-const reviewSubmitInitialView = (
+const mockUpdateReport = jest.fn();
+
+const mockedReportContext_InProgress = {
+  ...mockReportContext,
+  updateReport: mockUpdateReport,
+};
```
Comment (ntsummers1) [inline services/ui-src/src/routes/ReviewSubmit/ReviewSubmit.test.tsx:31]: I think if you delete this code...

File: services/ui-src/src/routes/ReviewSubmit/ReviewSubmit.test.tsx [hunk 3]
```diff
@@ -6,84 +6,73 @@ import { ReviewSubmit } from "routes";
...
+
+const ReviewSubmitComponent_InProgress = (
   <RouterWrappedComponent>
-    <ReportContext.Provider value={mockReportInitialContext}>
+    <ReportContext.Provider value={mockedReportContext_InProgress}>
```
Comment (ntsummers1) [inline services/ui-src/src/routes/ReviewSubmit/ReviewSubmit.test.tsx:35]: ... Swap this to
```suggestion
    <ReportContext.Provider value={mockReportContext}>
```

File: services/ui-src/src/routes/ReviewSubmit/ReviewSubmit.test.tsx [hunk 4]
```diff
@@ -92,25 +81,25 @@ describe("Test /mcpar/review-and-submit view", () => {
...
-  test("should fire the submitForm function when confirming the modal", async () => {
-    render(reviewSubmitInitialView);
+  test("ReviewSubmit updates report status on submit confirmation", async () => {
+    render(ReviewSubmitComponent_InProgress);
     const reviewSubmitButton = screen.getByText("Submit MCPAR")!;
     await userEvent.click(reviewSubmitButton);
     const modalSubmitButton = screen.getByTestId("modal-submit-button")!;
     await userEvent.click(modalSubmitButton);
-    await expect(mockReportMethods.updateReport).toHaveBeenCalledTimes(1);
+    await expect(mockUpdateReport).toHaveBeenCalledTimes(1);
```
Comment (ntsummers1) [inline services/ui-src/src/routes/ReviewSubmit/ReviewSubmit.test.tsx:90]: ... And change this to this
```suggestion
    await expect(mockUpdateReport.updateReport).toHaveBeenCalledTimes(1);
```

You can save yourself a few lines of code

**Review by ntsummers1 — APPROVED** [review]
Overall very nice! Just a few questions I had about if we can better use all the work you've done with the pulled out mockedReport code.

**Comment by gmrabian** [comment]
Noticed a bug with loading different reports. I think it's because the hydration in `Form.tsx` changed so it now only runs if there is reportData. I tested it without that condition, and it properly clears the field
https://user-images.githubusercontent.com/57802560/189674544-ebc9acf8-127d-4f77-bf48-e13e3dcb21e0.mov

**Review by gmrabian — CHANGES_REQUESTED** [review]
blocking on above bug

File: services/ui-src/src/components/forms/Form.tsx [hunk 5]
```diff
@@ -36,9 +36,12 @@ export const Form = ({ id, formJson, onSubmit, children, ...props }: Props) => {
...
   // hydrate and create form fields using formFieldFactory
-  const formFieldsToRender = (fields: FormField[]) => {
-    const hydratedFields = hydrateFormFields(fields, reportData);
-    return formFieldFactory(hydratedFields);
+  const renderFormFields = (fields: FormField[]) => {
+    let fieldsToRender = fields;
+    if (reportData) {
+      fieldsToRender = hydrateFormFields(fields, reportData);
+    }
+    return formFieldFactory(fieldsToRender);
```
Comment (ntsummers1) [inline services/ui-src/src/components/forms/Form.tsx:44]: To fix the bug Garrett pointed out where two reports and hydrate the others fields, you could always do something like this:

```suggestion
    const fieldsToRender = reportData
      ? hydrateFormFields(fields, reportData)
      : hydrateFormFields(fields, {});
    return formFieldFactory(fieldsToRender);
```

**Review by gmrabian — APPROVED** [review]

**Comment by ntsummers1** [comment]
Looks good to me!

**Pull request merged**